### ISSUE TITLE
Same memory consumption error independent of pandas training size

### ISSUE THREAD

**Issue by simpsus** [manujosephv/pytorch_tabular#176]
I have a weird error that I hope I can get some guidance on.
My train/val set are large, and I would like to iterate down towards a size that pytorch_tabular can run with. 
My problem is that no matter how small I make the training and validation dataset size, I get the exact same mnemory allocation error:

`  File "C:\Users\grumb\miniconda3\envs\numerai\lib\site-packages\pytorch_tabular\models\common\layers.py", line 155, in forward
    sim = einsum("b h i d, b h j d -> b h i j", q, k) * self.scale
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 2.78 GiB (GPU 0; 12.00 GiB total capacity; 8.58 GiB already allocated; 1.46 GiB free; 8.60 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF`

The numbers do not change whether or not I use my full dataset or just a portion of it. I even tried it with:
```
train.info()
<class 'pandas.core.frame.DataFrame'>
Index: 19703 entries, n000506a8e6b72b7 to nfff8d27687cbc80
Columns: 1210 entries, era to erano
dtypes: float32(1207), int32(1), object(2)
memory usage: 91.2+ MB

val.info()`
<class 'pandas.core.frame.DataFrame'>
Index: 25629 entries, n0006f5aa9a6275c to nffdad4502227c2f
Columns: 1210 entries, era to erano
dtypes: float32(1207), int32(1), object(2)
memory usage: 118.7+ MB
```
So, really small datasets that should be no trouble for my 12gb GPU.

I start the test with

```logger.debug("Creating TabularModel")
tabular_model = TabularModel(
    data_config=DataConfig(
        target=[target],
        continuous_cols=features,
        normalize_continuous_features=False,  # said to reduce memory usage
    ),
    model_config=FTTransformerConfig(
        task="regression",
        batch_norm_continuous_input=False,
    ),
    optimizer_config=OptimizerConfig(),
    trainer_config=TrainerConfig(auto_lr_find=False, max_epochs=100, accelerator="gpu"),
)
logger.debug("Starting Tabular Model fit")
tabular_model.fit(
    train=train,
    validation=val,
    loss=spearman,
)
```

Is there any hint/idea what could cause such a behavior?
Any tests that I can run to see which dataset/size was passed?

Thank you very much

**Comment by simpsus** [comment]
one more test is to substitute the data loading code with numpy random:

`train = pd.DataFrame(np.random.rand(100000,1000), columns=[str(i) for i in range(1000)])`

that works without any problems. So it has to do someting with my dataset. Though I have not idea what....

**Comment by manujosephv** [comment]
Having a smaller dataset wouldn't typically help with the GPU memory error. The GPU memory error is dependent on the `batch_size` that we use. Having a smaller dataset generally only helps with the RAM usage, 

So, maybe try reducing the `batch_size` in the `DataConfig` to fit it into GPU memory.

Also, looking at the sample train shape, I see you have 1000 features. Transformer based models will have large memory consumption because of the attention calculation. So, the `batch_size` would have to be really small.  Maybe you can try other, more efficient, models like GATE.

Disclaimer: I am the author of the GATE paper.

**Comment by simpsus** [comment]
Thank you very much for these insights. I would love to try the GATE model. Is this the [paper](https://arxiv.org/pdf/2207.08548.pdf)?
My dataset is completely numeric, all features are binned to 5 discrete values, dimensions ~2.5mio times 1500 but I could also select a smaller subset of the 1500 features. I would have several targets to train on, if one of the models available supports multiple targets.

Back to my memory issue. Thank you for the explanation on the `batch_size`. This restores my sanity about the dataset size not affecting RAM allocation. The thing is: I did neither set a batch_size nor did I change it in my tests between my dataset and the `np.random.rand()` data. So I am still at loss as to what causes it.

I find [here](https://pytorch-tabular.readthedocs.io/en/latest/apidocs_config/) that the default batch size is 64. That is already very low imho. I would my expect my 12GB GPU to handle a batch size of 64.

The following code gives me an out of memory error
```
import warnings
warnings.filterwarnings("ignore", category=UserWarning)
warnings.simplefilter(action="ignore", category=FutureWarning)
import torch
from loguru import logger
import numpy as np
import pandas as pd
from pytorch_tabular import TabularModel
from pytorch_tabular.models import FTTransformerConfig
from pytorch_tabular.config import (
    DataConfig,
    OptimizerConfig,
    TrainerConfig,
)

device = "cuda" if torch.cuda.is_available() else "cpu"
logger.debug(f"Device {device}")

train = pd.DataFrame(np.random.rand(100000,1000), columns=[str(i) for i in range(1000)]).astype(np.float32)
tabular_model = TabularModel(
    data_config=DataConfig(
        target=['999'],
        continuous_cols=[str(i) for i in range(999)],
        normalize_continuous_features=False,  # said to reduce memory usage
    ),
    model_config=FTTransformerConfig(
        task="regression",
        batch_norm_continuous_input=False,
    ),
    optimizer_config=OptimizerConfig(),
    trainer_config=TrainerConfig(auto_lr_find=True, max_epochs=100, accelerator="gpu"),
)
logger.debug("Starting Tabular Model fit")
tabular_model.fit(
    train=train,
    validation=train.copy(),
)
```
results in 
```
2023-03-12 20:16:51.514 | DEBUG    | __main__:<module>:17 - Device cuda
2023-03-12 20:16:52,266 - {pytorch_tabular.tabular_model:102} - INFO - Experiment Tracking is turned off
2023-03-12 20:16:52.374 | DEBUG    | __main__:<module>:33 - Starting Tabular Model fit
Global seed set to 42
2023-03-12 20:16:53,051 - {pytorch_tabular.tabular_model:465} - INFO - Preparing the DataLoaders
2023-03-12 20:16:53,754 - {pytorch_tabular.tabular_datamodule:286} - INFO - Setting up the datamodule for regression task
2023-03-12 20:16:53,819 - {pytorch_tabular.tabular_model:508} - INFO - Preparing the Model: FTTransformerModel
2023-03-12 20:16:53,977 - {pytorch_tabular.tabular_model:264} - INFO - Preparing the Trainer
Auto select gpus: [0]
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
2023-03-12 20:16:54,454 - {pytorch_tabular.tabular_model:558} - INFO - Auto LR Find Started
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
<-- a lot of stacktrace -->
OutOfMemoryError: CUDA out of memory. Tried to allocate 1.91 GiB (GPU 0; 12.00 GiB total capacity; 9.74 GiB already allocated; 219.42 MiB free; 9.77 GiB reserved in total by PyTorch)
```

If I change the batch_size in the trainer config to 32
```
trainer_config=TrainerConfig(
        batch_size=32, auto_lr_find=True, max_epochs=100, accelerator="gpu"
    )
```
I get closer but still an out of memory error
`OutOfMemoryError: CUDA out of memory. Tried to allocate 978.00 MiB (GPU 0; 12.00 GiB total capacity; 10.83 GiB already allocated; 0 bytes free; 11.02 GiB reserved in total by PyTorch)`
Changing it to `16` .. it runs!
It runs at an atrocious 3it/s and one hour per epoch, but it runs.

Switching the FFTConfig for a GATE config:
```
model_config=GatedAdditiveTreeEnsembleConfig(
        task="regression",
    )
```
it runs even slower at 0.8 it/s.

To summarize:
1. Thank you!
2. Is 16 the maximum batch size I can achieve?
3. Do you have any hints on which settings could make the GATE model faster?
4. I want to control the batches further. I have feature (not used for training) that groups samples (~5k sample groups). Ideally, I would want to use those for batches. Given my current 16, that will remain a dream. My question: How can I write a custome sampler so that I can control that the 16 provided samples are from the same group?

**Comment by manujosephv** [comment]
1. Yes, you got the paper right.
2. FTTransformer uses a self attention calculation across features and that operation has an O(N^2) memory complexity, if I'm not mistaken. So with 1500+ features, FTTransformer is going to take a lot of memory. So, having to use a really small batch size does make sense here.
3. GATE model is actually just research code and not really optimized for performance. One way to speed up is to reduce the depth of the trees and the number of trees. Turning off chaining may also speed up the model. All these would have impacts on how good the model does as well.
4. PyTorch Tabular supports native PyTorch Batch Samplers. You can pass them i nthe fit methood using the parameter `train_sampler`
